### PR TITLE
Use Trace Sampling Rate as configuration with probabilistic trace sampler 

### PR DESCRIPTION
**Pull request by shalabhms**
# Description
- Adding the Tracing Sampling Rate Config in the TracingSpec. Removed "Enabled" property.
- Sampling Rate 0 signifies, tracing is disabled because sampler will turned off sampling. Proababilistic Sampler is used with configured sampling rate.
- Using the Default Sampling Rate as 0.001 
- Adding the trace utils to have trace helper methods.
- Changes to gRPC and HTTP server and API packages.

## Issue reference

#1395 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation
* [ ] Specification has been updated
* [ ] Provided sample for the feature

### CODE REVIEW

File: pkg/apis/configuration/v1alpha1/types.go [hunk 1]
```diff
@@ -64,9 +64,7 @@ type SelectorField struct {
 
 // TracingSpec is the spec object in ConfigurationSpec
 type TracingSpec struct {
-	Enabled      bool `json:"enabled"`
-	IncludeBody  bool `json:"includeBody"`
-	ExpandParams bool `json:"expandParams"`
+	SamplingRate string `json:"samplingRate"`
```
Comment (youngbupark) [inline pkg/apis/configuration/v1alpha1/types.go:67]: did you regen the crd clients?

Comment (shalabhms) [reply pkg/apis/configuration/v1alpha1/types.go:67]: > did you regen the crd clients?

No, that is pending, I mentioned in the PR description. I tried regenerating and script succeeded , as it did not complain but there are no files update actually happening. So I am checking and might need help.

File: pkg/diagnostics/tracing.go [hunk 2]
```diff
@@ -216,13 +223,16 @@ func TracingSpanFromGRPCContext(c context.Context, req interface{}, method strin
...
 	}
 
+	rate := diag_utils.GetTraceSamplingRate(spec.SamplingRate)
+
+	// TODO : Continue using ProbabilitySampler till Go SDK starts supporting RateLimiting sampler
 	if corID != "" {
 		spanContext := DeserializeSpanContext(corID)
-		ctx, span = trace.StartSpanWithRemoteParent(c, method, spanContext, trace.WithSpanKind(trace.SpanKindServer))
-		ctxc, spanc = trace.StartSpanWithRemoteParent(ctx, createSpanName(method), span.SpanContext(), trace.WithSpanKind(trace.SpanKindClient))
+		ctx, span = trace.StartSpanWithRemoteParent(c, method, spanContext, trace.WithSpanKind(trace.SpanKindServer), trace.WithSampler(trace.ProbabilitySampler(rate)))
```
[the remark this answers is not in the document]
Comment (shalabhms) [reply pkg/diagnostics/tracing.go]: Done and changed others as well in the call.

File: pkg/apis/configuration/v1alpha1/types.go (hunk 1 above)
Comment (shalabhms) [reply pkg/apis/configuration/v1alpha1/types.go:67]: as discussed with @youngbupark  , TracingSpec struct is deep copied and not through memberwise copy, so crd update is not needed.

File: pkg/diagnostics/utils/trace_utils.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// ------------------------------------------------------------
+
+package utils
+
+import (
+	"strconv"
+)
+
+const (
+	defaultTraceSamplingRate = 0.001
```
Comment (youngbupark) [inline pkg/diagnostics/utils/trace_utils.go:13]: Can we use 1e-4 to follow opencensus default rate - https://github.com/census-instrumentation/opencensus-go/blob/46dfec7deb6e8c5d4a46f355c0da7c6d6dc59ba4/trace/sampling.go#L21 ?

Comment (shalabhms) [reply pkg/diagnostics/utils/trace_utils.go]: done.

File: pkg/diagnostics/tracing.go [hunk 4]
```diff
@@ -77,13 +78,20 @@ func TraceSpanFromFastHTTPRequest(r *fasthttp.Request, spec config.TracingSpec)
...
 		spanContext := DeserializeSpanContext(corID)
-		ctx, span = trace.StartSpanWithRemoteParent(context.Background(), uriSpanName, spanContext, trace.WithSpanKind(trace.SpanKindServer))
-		ctxc, spanc = trace.StartSpanWithRemoteParent(ctx, createSpanName(uriSpanName), span.SpanContext(), trace.WithSpanKind(trace.SpanKindClient))
+		ctx, span = trace.StartSpanWithRemoteParent(context.Background(), uriSpanName, spanContext, serverKindOption, probSamplerOption)
+		ctxc, spanc = trace.StartSpanWithRemoteParent(ctx, spanName, span.SpanContext(), clientKindOption, probSamplerOption)
 	} else {
-		ctx, span = trace.StartSpan(context.Background(), uriSpanName, trace.WithSpanKind(trace.SpanKindServer))
-		ctxc, spanc = trace.StartSpanWithRemoteParent(ctx, createSpanName(uriSpanName), span.SpanContext(), trace.WithSpanKind(trace.SpanKindClient))
+		ctx, span = trace.StartSpan(context.Background(), uriSpanName, serverKindOption, probSamplerOption)
+		ctxc, spanc = trace.StartSpanWithRemoteParent(ctx, spanName, span.SpanContext(), clientKindOption, probSamplerOption)
```
Comment (amanbha) [inline pkg/diagnostics/tracing.go:94]: This block seems too repetitive, can be moved to a method

File: pkg/diagnostics/utils/trace_utils.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	"strconv"
+)
+
+const (
+	defaultSamplingRate = 1e-4
+)
+
+// GetTraceSamplingRate parses the given rate and returns the parsed rate
+func GetTraceSamplingRate(rate string) float64 {
+	f, err := strconv.ParseFloat(rate, 64)
```
[the remark this answers is not in the document]
Comment (shalabhms) [reply pkg/diagnostics/utils/trace_utils.go:18]: It is validated in sampler code : 

https://github.com/census-instrumentation/opencensus-go/blob/46dfec7deb6e8c5d4a46f355c0da7c6d6dc59ba4/trace/sampling.go#L46

File: pkg/diagnostics/tracing.go (hunk 4 above)
Comment (shalabhms) [reply pkg/diagnostics/tracing.go]: Refactored to getTraceSpan method for http requests. However leaving out gRPC one as it is only one method that creating tracespan, as parent /child span tight, so keeping that one intact.

**Comment by yaron2** [comment]
Tests are broken.

**Comment by shalabhms** [comment]
> Tests are broken.

Failed Test suite have passed at my local cluster. Looking at it though...

**Pull request merged**